### PR TITLE
Integration of RAFT Optical Flow model to SG.

### PR DESCRIPTION
**Pull request by Yael-Baron**

### CODE REVIEW

File: src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        coords0 = self.coords_grid(N, H // 8, W // 8, device=img.device)
+        coords1 = self.coords_grid(N, H // 8, W // 8, device=img.device)
+
+        # optical flow computed as difference: flow = coords1 - coords0
+        return coords0, coords1
+
+    def forward(self, x, **kwargs):
+        """Estimate optical flow between pairs of frames"""
+
+        image1 = x[:, 0]
```
Comment (BloodAxe) [inline src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py:68]: 1) The way of passing input image pair as [B,2,C,H,W] should be reflected in docstring
2) No kwargs

File: src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        # run the context network
+        net, inp = self.context_encoder(image1)
+
+        # initialize flow
+        coords0, coords1 = self.initialize_flow(image1)
+
+        # run update block network
+        flow_predictions, flow_up = self.flow_iterative_block(coords0, coords1, net, inp, fmap1, fmap2)
+
+        if not self.training:
```
Comment (BloodAxe) [inline src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py:83]: Why we need this?
I suggest the model to have same output format regardless whether it is in training/eval mode.
If you want to omit some outputs when exporting model to onnx you can do `if torch.jit.is_tracing()` but I suggest not to make output of the model be state-dependent.

File: tests/unit_tests/export_optical_flow_model_test.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+class TestOpticalFlowModelExport(unittest.TestCase):
+    def setUp(self) -> None:
+        logging.getLogger().setLevel(logging.DEBUG)
+
+        self.models_to_test = [
+            Models.RAFT_S,
+            Models.RAFT_L,
+        ]
+
+    # def test_infer_input_image_shape_from_model(self):
```
Comment (BloodAxe) [inline tests/unit_tests/export_optical_flow_model_test.py:29]: Please remove irrelevant commented code pieces

**Comment by BloodAxe** [comment]
I have a few questions regarding this PR:
- Are we going to include pretrained weights for RAFT? (Official or ours)?
- What about dataset & training recipes & loss? Without it it sort of does not make any sense to include a model.

File: src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+from super_gradients.common.registry import register_model
+from super_gradients.module_interfaces import SupportsReplaceInputChannels
+from super_gradients.module_interfaces.exportable_optical_flow import ExportableOpticalFlowModel
+from super_gradients.training.models import get_arch_params, SgModule
+from super_gradients.training.utils.utils import HpmStruct, get_param
+
+from .raft_base import Encoder, ContextEncoder, FlowIterativeBlock
+
+
+class RAFT(ExportableOpticalFlowModel, SgModule):
```
Comment (BloodAxe) [inline src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py:19]: Please include a docstring with a link to original paper/implementation

File: src/super_gradients/module_interfaces/exportable_optical_flow.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+            # For FP16 quantization, we simply can to convert the whole model to half precision
+            complete_model = complete_model.half()
+
+            if calibration_loader is not None:
+                logger.warning(
+                    "It seems you've passed calibration_loader to export function, but quantization_mode is set to FP16. "
+                    "FP16 quantization is done by calling model.half() so you don't need to pass calibration_loader, as it will be ignored."
+                )
+
+        if engine in {ExportTargetBackend.ONNXRUNTIME, ExportTargetBackend.TENSORRT}:
```
Comment (shaydeci) [inline src/super_gradients/module_interfaces/exportable_optical_flow.py:226]: I thought we didnt check that TRT conversion works.
We should not add code that has not been tested.

File: src/super_gradients/module_interfaces/exportable_optical_flow.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            )
+        elif quantization_mode == ExportQuantizationMode.FP16:
+            if contains_quantized_modules:
+                raise RuntimeError("Model contains quantized modules for INT8 mode. " "FP16 quantization is not supported for such models.")
+        elif quantization_mode is None and contains_quantized_modules:
+            # If quantization_mode is None, but we have quantized modules in the model, we need to
+            # update the quantization_mode to INT8, so that we can correctly export the model.
+            quantization_mode = ExportQuantizationMode.INT8
+
+        from super_gradients.training.models.conversion import ConvertableCompletePipelineModel
```
Comment (shaydeci) [inline src/super_gradients/module_interfaces/exportable_optical_flow.py:208]: Why is the import here ?

File: src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            hidden_dim=encoder_params.hidden_dim,
+            context_dim=encoder_params.context_dim,
+            output_dim=encoder_params.cnet.output_dim,
+            norm_fn=encoder_params.cnet.norm_fn,
+            dropout=encoder_params.dropout,
+        )
+
+        self.flow_iterative_block = FlowIterativeBlock(encoder_params, encoder_params.update_block.hidden_dim, flow_params, corr_params.alternate_corr)
+
+    def freeze_bn(self):
```
Comment (shaydeci) [inline src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py:45]: What's the use of this ?

File: src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py (hunk 4 above)
Comment (shaydeci) [reply src/super_gradients/training/models/optical_flow_models/raft/raft_variants.py:19]: And add docstrings to the rest of the methods of this class.

**Review by shaydeci — CHANGES_REQUESTED** [review]

**Comment by Yael-Baron** [comment]
@BloodAxe 

1. We're planing to add pre-trained weights for RAFT_L and RAFT_S.
2. There will be a PR for dataset, transforms, loos+metric separately.

**Comment by Yael-Baron** [comment]
See PR Feature/dle 599 optical flow integration #1984.
